### PR TITLE
node-neigh: Fix node removal and invalid neigh entry due to buggy arping response correlation

### LINKED ISSUE #14693: cilium 1.9.2 ARP errors
After update to cilium 1.9.2 I started to observe these errors in my ARP tables on k8s masters:

```
kube-5.prod.xxx: arp -a|grep "xx:xx:8c:ae:ef:43"
kube-3.prod.xxx (xxx.xxx.173.56) at xx:xx:8c:ae:ef:43 [ether] PERM on eth0
 
kube-3.prod.xxx: arp -a|grep "xx:xx:8c:ae:ef:43"
kube-2.prod.xxx (xxx.xxx.173.55) at xx:xx:8c:ae:ef:43 [ether] PERM on eth0
```

As you can see on different nodes the same MAC address is bound to different IPv4 addresses.

Bare-metal K8s:
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.5", GitCommit:"e338cf2c6d297aa603b50ad3a301f761b4173aa6", GitTreeState:"clean", BuildDate:"2020-12-09T11:18:51Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.5", GitCommit:"e338cf2c6d297aa603b50ad3a301f761b4173aa6", GitTreeState:"clean", BuildDate:"2020-12-09T11:10:32Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```

P.S. Just noticed that my kube-proxy DS was up and running. Maybe related.

### PR DESCRIPTION
**Pull request by brb**
Please see commit messages.

For backporters: please ignore the last commit.

Fixes #14693.

### CODE REVIEW

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-net-next

**Comment by brb** [comment]
test-4.9

**Comment by brb** [comment]
test-gke

**Comment by brb** [comment]
test-gke

**Comment by brb** [comment]
test-me-please

File: pkg/datapath/linux/node.go [hunk 1]
```diff
@@ -609,11 +609,45 @@ func (n *linuxNodeHandler) insertNeighbor(newNode *nodeTypes.Node, ifaceName str
...
+			neigh, found := n.neighByNextHop[existingNextHopStr]
+			if found {
+				if err := netlink.NeighDel(neigh); err != nil {
+					log.WithFields(logrus.Fields{
+						logfields.IPAddr:       neigh.IP,
+						logfields.HardwareAddr: neigh.HardwareAddr,
+						logfields.LinkIndex:    neigh.LinkIndex,
+					}).WithError(err).Warn("Failed to remove neighbor entry")
+				}
+				delete(n.neighByNextHop, nextHopStr)
```
Comment (aanm) [inline pkg/datapath/linux/node.go:647]: Should we still delete it even if the `netlink` returned an error?

File: pkg/datapath/linux/node.go [hunk 2]
```diff
@@ -588,7 +588,7 @@ func (n *linuxNodeHandler) getSrcAndNextHopIPv4(nodeIPv4 net.IP, ifaceName strin
 }
 
 // Must be called with linuxNodeHandler.mutex held.
-func (n *linuxNodeHandler) insertNeighbor(newNode *nodeTypes.Node, ifaceName string) {
+func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeTypes.Node, ifaceName string, refresh bool) {
```
Comment (aanm) [inline pkg/datapath/linux/node.go:601]: Please add documentation to the parameters of this function, it's not clear to me anymore what are certain parameters used for.

**Review by aanm — COMMENTED** [review]
Some small nits, doesn't require my 2nd review.

File: pkg/datapath/linux/node.go (hunk 1 above)
Comment (brb) [reply pkg/datapath/linux/node.go:647]: Yeah, as otherwise nobody would remove it from the map.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-net-next

**Review by aanm — APPROVED** [review]

File: pkg/datapath/linux/arp/ping.go [hunk 3]
```diff
@@ -98,7 +98,8 @@ func (p *pinger) resolve(ip net.IP) (net.HardwareAddr, error) {
 			return nil, err
 		}
 
-		if resp.Operation != layers.ARPReply || !bytes.Equal(resp.SourceProtAddress, ip.To4()) {
+		if resp.Operation != layers.ARPReply || !bytes.Equal(resp.SourceProtAddress, ip.To4()) ||
```
Comment (rbtcollins) [inline pkg/datapath/linux/arp/ping.go]: Applying deMorgan here would make this a little easier to read:

```
if resp.Operation == layers.ARPReply && bytes.Equal(resp.SourceProtAddress, ip.To4()) &&
bytes.Equal(resp.DstProtAddress, p.ip.To4()) {
```

**Comment by brb** [comment]
test-net-next

File: pkg/datapath/linux/node_linux_test.go [hunk 4]
```diff
@@ -1025,6 +1025,11 @@ func (s *linuxPrivilegedIPv4OnlyTestSuite) TestArpPingHandling(c *check.C) {
 	}
 	err = linuxNodeHandler.NodeAdd(nodev1)
 	c.Assert(err, check.IsNil)
+	// Insert the same node second time. This should not increment refcount for
+	// the same nextHop. We test it when check that NodeDelete has removed the
```
Comment (pchaigno) [inline pkg/datapath/linux/node_linux_test.go]: Nit: missing word in second sentence.

File: pkg/datapath/linux/node.go [hunk 5]
```diff
@@ -562,11 +562,11 @@ func (n *linuxNodeHandler) getSrcAndNextHopIPv4(nodeIPv4 net.IP, ifaceName strin
 	// Figure out whether nodeIPv4 is directly reachable (i.e. in the same L2)
 	routes, err := netlink.RouteGet(nodeIPv4)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to retrieve route for remote node IP: %w", err)
+		return nil, nil, fmt.Errorf("Failed to retrieve route for remote node IP: %w", err)
 	}
 
 	if len(routes) == 0 {
-		return nil, nil, fmt.Errorf("remote node IP is non-routable")
+		return nil, nil, fmt.Errorf("Remote node IP is not routable. Connectivity to pods on that node may be unavailable.")
```
Comment (pchaigno) [inline pkg/datapath/linux/node.go]: I think this commit is reverting all changes from the two previous commits by mistake.

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: pkg/datapath/linux/node.go (hunk 5 above)
Comment (brb) [reply pkg/datapath/linux/node.go]: :facepalm:

**Comment by brb** [comment]
test-me-please

File: pkg/datapath/linux/node.go (hunk 5 above)
Comment (pchaigno) [reply pkg/datapath/linux/node.go]: I don't think this is resolved. Most of commit `node-neigh: Unconditionally remove neigh entry from BPF map` is just reverts of the two previous commits.

Comment (brb) [reply pkg/datapath/linux/node.go]: :facepalm: x2 (having two workdirs doesn't help)

**Comment by brb** [comment]
test-me-please

**Review by pchaigno — APPROVED** [review]

**Pull request merged**